ccl: make tests work with a tenant

andreimatei opened the pull request:
This PR makes all tests under `pkg/ccl` compatible with multi-tenancy - i.e. compatible with the `COCKROACH_TEST_TENANT_MODE=forceTenant` mode whereby all `TestServers` auto-magically get a secondary tenant.

When the respective env var is not set, each test has a 50% chance of running in this mode if some of the `ccl` packages were linked in, and if the tests (or the test package) installed a testing license. Currently, a bunch of the ccl tests are not compatible with this mode; however, then problem was silently hidden because these tests were being skipped [1] half the time (the multi-tenant half) because they we're linking in all of the `ccl` packages. This patch does not change the skipping part, that will come separately. However, this patch does make the tests pass instead of fail, where they not to be skipped.


[1] https://github.com/cockroachdb/cockroach/blob/93d78cbcb5b196bdee7a71d3a5a8ce7774b751e6/pkg/testutils/serverutils/test_server_shim.go#L80

Release note: None
Epic: None

andreimatei committed:
partitionccl,testutils: make partitionccl tests work with a tenant

This patch makes all tests under partitionccl work when
serverutils.StartServer() decides to start a tenant inside the server
that it returns. StartServer does that randomly when
IsEnterpriseEnabled()==true (which it does because the main_test.go in
this pkg calls TestingEnableEnterprise).
Before this patch, when the random coin decided to add a tenant, the
tests ended up being silently skipped [1], because the binary with the
tests does not include all of the ccl packages. This skipping is bad.

The patch inhibits the creation of the tenant explicitly for a
few tests (where syntax unsupported in multi-tenancy is used - like zone
config stuff), and makes others more resilient.

The tests will continue to be skipped according to the coin like they
were before; this patch does not make them import all of ccl (but if
they did import all of ccl, they would now pass instead of fail). All of
ccl will likely be linked in the future by these tests (I got here by
working on such a change), but I'll do that separately.

Many other ccl subpackages probably need similar treatment; this pkg was
chosen arbitrarily to start with and see how it goes.

Tested with
dt pkg/ccl/partitionccl --stress -- --test_env="COCKROACH_TEST_TENANT_MODE=forceTenant"
and with adding
import _ "github.com/cockroachdb/cockroach/pkg/ccl"
to a file in the `partitionccl_test` pkg.

[1] https://github.com/cockroachdb/cockroach/blob/93d78cbcb5b196bdee7a71d3a5a8ce7774b751e6/pkg/testutils/serverutils/test_server_shim.go#L80

Release note: None
Epic: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/92024)
<!-- Reviewable:end -->

andreimatei committed:
oidcccl: move TestMain to main_test.go

... as is customary. main_test.go comes with the advantage that it can
be in the _test package, which I'll use later to have it import all
"ccl".

Release note: None

andreimatei committed:
oidcccl: make tests compatible with multi-tenancy

When multi-tenancy was enabled, these tests were executing SQL in one
tenant, but then attempting to login into the system tenant. This
problem was hidden because the tests are skipped dynamically when
multi-tenancy is enabled. This patch does not change the skipping, but
that will be done later when I start importing "ccl" from all the ccl
tests.

Release note: None
Epic: None

andreimatei committed:
changefeedccl/kvfeed: make tests compatible with multi-tenancy

When multi-tenancy was enabled, these tests were executing SQL in one
tenant, but then attempting to login into the system tenant. This
problem was hidden because the tests are skipped dynamically when
multi-tenancy is enabled. This patch does not change the skipping, but
that will be done later when I start importing "ccl" from all the ccl
tests.

The patch makes the test work with a tenant, though, by not hardcoding
the use of the system codec.

Release note: None
Epic: None

ajstorm reviewed (changes requested):
> Reviewed 5 of 7 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/92024)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andreimatei, @herkolategan, and @smg260)

___
*[`pkg/ccl/partitionccl/scrub_test.go` line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/92024#-NH6ZvI09f3vNpvv3Zcu:-NH6ZvI09f3vNpvv3Zcv:b-murgt7) ([raw file](https://github.com/cockroachdb/cockroach/blob/28f01788f44c524a4d73e8e3696fa8c598bed7be/pkg/ccl/partitionccl/scrub_test.go#L30)):*
> ```Go
> )
> 
> func systemOrFirstTenantCodec(s serverutils.TestServerInterface) keys.SQLCodec {
> ```

extreme nit: the naming of this function implies to me that we try to get the system tenant's codec and if not, we get the first tenant's codec.  That's not the case though, as we only get the system tenant's codec if secondary tenant doesn't exist.  Additionally, it's really a system or tenant codec, so whether or not it's the first tenant doesn't really matter.  It might be more accurate to name it something like `tenantOrSystemCodec()`, or maybe even `getConnectionCodec()` to imply that the codec is the one that belongs to the current connection that we're using.

Also, less nitty: Should this be in `serverutils` as a function off of the `TestServerInterface` directly?  I could imagine others making use of this down the road.


<!-- Sent from Reviewable.io -->

andreimatei committed:
authccl: make tests compatible with multi-tenancy

When multi-tenancy was enabled, these tests were doing some things in
the system tenant, some things in the secondary tenant, and thus they
were broken. This problem was hidden because the tests are skipped
dynamically when multi-tenancy is enabled. This patch does not change
the skipping, but that will be done later when I start importing "ccl"
from all the ccl tests.

The patch makes the tests work with a tenant, when there is one.

Release note: None

andreimatei reviewed (commented):
A few more commits now. I believe this now makes all of the ccl tests pass in multi-tenant mode.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/92024#-:-NH71g6--YGW2hQ80BqQ:b5cedcy)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @ajwerner, @herkolategan, and @smg260)

___
*[`pkg/ccl/partitionccl/scrub_test.go` line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/92024#-NH6ZvI09f3vNpvv3Zcu:-NH6egQ7Ejob5TM-3q6X:bu4tzam) ([raw file](https://github.com/cockroachdb/cockroach/blob/28f01788f44c524a4d73e8e3696fa8c598bed7be/pkg/ccl/partitionccl/scrub_test.go#L30)):*
> or maybe even getConnectionCodec() to imply that the codec is the one that belongs to the current connection that we're using.


From the perspective of this function, there is no "connection that we're using" (and therein lies the larger tragedy, in my opinion, with `StartServer` returning a connection pool that points to either one tenant or another, nondeterministically; the connection that one is using is not clear).

> Additionally, it's really a system or tenant codec, so whether or not it's the first tenant doesn't really matter.

Is that true? Different tenants (if we were to have many of them) have different codecs, don't they? So "the first" part is relevant, I think. Still - 

I've renamed to `tenantOrSystemCodec()`.

---

> Also, less nitty: Should this be in serverutils as a function off of the TestServerInterface directly? I could imagine others making use of this down the road.

I've considered it, but opted to keep this function hidden because I don't like it at all. If the `TestServer` is to sometimes create a tenant for you, I think we should figure out another set of return values to `StartServer` (or deprecate that function and start over), such that you can get the rid codec directly from the `s` it returns (i.e. the codec matching the tenant that the connection pool is tied to).
I see that this kind of magic "if there's one tenant, do this, if there's none, do that, if there's more than one freak out" logic was transparently added to some `TestServer` methods (in particular `ServingSQLAddr`), but I for one think it's a bad idea.

I did have to add this function to one more packages, but that seems to be the extent of the current problem; I prefer having two copies of it. For other tests I've had to add other similar functions for other `TestServer` fields and such, so I wouldn't even know where to stop adding things to `TestServer`. I do think that a half-magic state where some methods looks at the tenant and others don't wouldn't not be good.


<!-- Sent from Reviewable.io -->

ajstorm reviewed (changes requested):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

We can chat more on Monday about how to proceed on some of the more fundamental reorganization.

> Reviewed 1 of 7 files at r1, 6 of 6 files at r4, 3 of 3 files at r5, 2 of 2 files at r6, 3 of 3 files at r7, 1 of 1 files at r8, 2 of 2 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/92024#-:-NHAaM4e1R4SztGp39qk:b-9sw8zk)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @andreimatei, @herkolategan, and @smg260)

___
*[`pkg/ccl/partitionccl/scrub_test.go` line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/92024#-NH6ZvI09f3vNpvv3Zcu:-NHAXIbWAuA-gzwE3thU:bcvq9jh) ([raw file](https://github.com/cockroachdb/cockroach/blob/28f01788f44c524a4d73e8e3696fa8c598bed7be/pkg/ccl/partitionccl/scrub_test.go#L30)):*
> Different tenants (if we were to have many of them) have different codecs, don't they?

Yes, but since we only ever have one tenant, it's either the test tenant or the system tenant.  It was probably a bad idea when I made this original change to make TestTenants a slice.  It would have been far more understandable if it were a singleton test tenant.  

> but I for one think it's a bad idea.

Yes, I agree.  Let's chat on Monday about a way forward here.  For the sake of this review, I think things are good enough for now.  


<!-- Sent from Reviewable.io -->

ajstorm reviewed (approved):
this time with the approval

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/92024#-:-NHAaU_M3oJzHocUJ5iU:bdzawqe)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @andreimatei, @herkolategan, and @smg260)



<!-- Sent from Reviewable.io -->

andreimatei commented:
TFTR!

bors r+

knz reviewed (commented):
thanks

> Reviewed 4 of 7 files at r1, 3 of 6 files at r4, 8 of 8 files at r10, 3 of 3 files at r11, 2 of 2 files at r12, 3 of 3 files at r13, 2 of 2 files at r14, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/92024#-:-NHOs7XJ0RBuL4BHkC-H:b-k73z76)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->
